src/nhs_faces.jl

```diff
@@ -0,0 +1,200 @@
+    empty_vector      :: Vector{Int} # Just an empty vector (used in `eachneighbor`)
+    n_cells           :: NTuple{NDIMS, Int}
+    periodic_box      :: Nothing
+    search_radius     :: ELTYPE
+
+    function FaceNeighborhoodSearch{NDIMS}(search_radius) where {NDIMS}
+        ELTYPE = eltype(search_radius)
+
+        cell_list = Dict{NTuple{NDIMS, Int}, Vector{Int}}()
+        neighbor_iterator = Dict{NTuple{NDIMS, Int}, Vector{Int}}()
+
+        cell_size = ntuple(dim -> search_radius, NDIMS)
+        empty_vector = Int[]
+
+        n_cells = ntuple(_ -> -1, Val(NDIMS))
+
+        new{NDIMS, ELTYPE, Nothing}(cell_list, cell_size, neighbor_iterator, empty_vector,
+                                    n_cells, nothing, search_radius)
+    end
+end
+
+@inline Base.ndims(::FaceNeighborhoodSearch{NDIMS}) where {NDIMS} = NDIMS
+
+@inline function eachneighbor(coords, neighborhood_search::FaceNeighborhoodSearch)
+    (; neighbor_iterator, empty_vector) = neighborhood_search
+    cell = cell_coords(coords, neighborhood_search)
+
+    haskey(neighbor_iterator, cell) && return neighbor_iterator[cell]
+
+    return empty_vector
+end
+
+function faces_in_cell(cell, neighborhood_search)
+    (; cell_list, empty_vector) = neighborhood_search
+
+    haskey(cell_list, cell) && return cell_list[cell]
+
+    return empty_vector
+end
+
+function initialize!(neighborhood_search::FaceNeighborhoodSearch, geometry;
+                     pad = ntuple(_ -> 1, ndims(geometry)))
+    (; cell_list, neighbor_iterator, search_radius) = neighborhood_search
+
+    empty!(cell_list)
+
+    # Fill cells with intersecting faces
+    for face in eachface(geometry)
+
+        # Check if any face intersects a cell in the face-embedding cell grid
+        for cell in cell_grid(face, geometry, neighborhood_search)
+            if cell_intersection(face, geometry, cell, neighborhood_search)
+                if haskey(cell_list, cell) && !(face in cell_list[cell])
+                    # Add face to corresponding cell
+                    append!(cell_list[cell], face)
+                else
+                    # Create cell
+                    cell_list[cell] = [face]
+                end
+            end
+        end
+    end
+
+    empty!(neighbor_iterator)
+
+    min_cell = cell_coords(geometry.min_corner, neighborhood_search) .- pad
+    max_cell = cell_coords(geometry.max_corner, neighborhood_search) .+ pad
+
+    # Merge all lists of faces in the neighboring cells into one iterator
+    face_ids = Int[]
+    for cell_runner in meshgrid(min_cell, max_cell)
+        resize!(face_ids, 0)
+        for neighbor in neighboring_cells(cell_runner, neighborhood_search,
+                                          search_radius)
+            append!(face_ids, faces_in_cell(Tuple(neighbor), neighborhood_search))
+        end
+
+        unique!(face_ids)
+
+        if isempty(face_ids)
+            continue
+        end
+
+        neighbor_iterator[cell_runner] = copy(face_ids)
+    end
```

Comment:
Why do you need to explicitly construct these lists compared to just looping over neighboring cells on the fly like the GNHS? Because faces can be in multiple cells, so you need to make it unique somehow?

Reply:
Yes, exactly. Otherwise, I'll loop over the same face multiple times in worse case.

Reply:
Hmmm, then the best idea I have is to use a second cell list to store the neighbors of each cell. This will be more flexible and probably more efficient, but you need to do the `unique!` part manually, I'm afraid, for it to work with `FullGridCellList`.

For a first implementation, you can hardcode a `DictionaryCellList` here.